### DISCUSSION TITLE
Implement JWT in Mirth

### DISCUSSION THREAD

**Discussion by 5011-5011** [nextgenhealthcare/connect#5503] [Q&A] [answered]
Hello! I would like to know if they have implemented JSON WEB TOKEN in Mirth. It's possible? If so, I appreciate any guidance. Thanks.

**Comment by pacmano1** [comment]
Yes possible, you have to code it, no easy button.

**Reply by 5011-5011** [reply]
Okay. Could you guide me if I should encode it in a filter or a transformer, or directly in the authentication section in the Source tab.
Thank u

**Reply by 5011-5011** [reply]
Okay. Could you guide me if I should encode it in a filter or a transformer, or directly in the authentication section in the Source tab.
Thank u

**Reply by pacmano1** [reply]
https://forums.mirthproject.io/forum/mirth-connect/support/18015-json-web-token-jwt-http-listener-authentication

**Reply by pacmano1** [reply]
Take a look around the forums.  I am not sure someone has ever posted a full solution.

**Reply by 5011-5011** [reply]
Thank you very much for always responding to my posts. I hope one day when I have learned enough to also share my knowledge with professionals who are starting at MIRTH. Thank you @pacmano1  :)

**Reply by pacmano1** [reply]
@5011-5011 , can you mark @jcurry5 's answer as the answer?  My stuff was a suggestion, his answer, if you confirm it works, is a better post.   Do post your version if you get it working to help others!

**Comment by jcurry5** [comment]
Here's how I did it:

Create an OAuth channel that runs every 4 minutes to generate a new JWT

Destination tab:  
I have a query parameter for 'client_assertation' that has a variable of ${JWT}
Create a new transformer as a JavaScript type
Code:
`var encoded;
var keySpec;
var claims = new Packages.org.jose4j.jwt.JwtClaims();
var jws = new Packages.org.jose4j.jws.JsonWebSignature();
var keyFac = java.security.KeyFactory.getInstance("RSA");
var privateKey;

encoded = Packages.org.apache.commons.codec.binary.Base64.decodeBase64("<base64 decoded from certificate>");
keySpec = new Packages.java.security.spec.PKCS8EncodedKeySpec(encoded);
privateKey = keyFac.generatePrivate(keySpec);

claims.setGeneratedJwtId();
claims.setIssuer('<put in issuer>');
claims.setSubject('<put in subject>');

claims.setNotBeforeMinutesInThePast(1);    
claims.setExpirationTimeMinutesInTheFuture(4);
claims.setIssuedAtToNow();
claims.setAudience("<url oauth>");

jws.setHeader('typ', 'JWT');
jws.setKey(privateKey);
jws.setPayload(claims.toJson());
jws.setAlgorithmHeaderValue(org.jose4j.jws.AlgorithmIdentifiers.RSA_USING_SHA384);

globalMap.put('JWT', jws.getCompactSerialization());`

I had to install the jose4j security library.

Create a new Response:
JavaScript type
Code:
`var access_token = msg['access_token'];
globalMap.put('access_token',access_token);`

In a different channel that performs the heavy lifting, I have a destination with an Authorization query parameter that references the access_token global variable:
Bearer ${access_token}

I learned this from very limited knowledge and after quite a bit of testing with a vendor this has worked pretty well.  But I hope this helps you on your way and what you're looking for.

**Reply by 5011-5011** [reply]
Thank you very much for sharing that information with me. Really :)

**Reply by 5011-5011** [reply]
@jcurry5  I add the security library in this section? Or do I install it via console? Thanks.

![image](https://user-images.githubusercontent.com/98954172/202709335-de2594c6-274a-4588-99fe-267a795da5ec.png)

**Reply by pacmano1** [reply]
That is the correct place - but be sure to allow it to be included in your channel on the right hand side of the code template editor.

i.e. if you highlight Library 2 in your use case you will see on the lower right the channels allowed to use it.

![image](https://user-images.githubusercontent.com/44065187/202714371-5fa027a4-b2d2-4652-b31a-d91cdd7b7804.png)

If you highlight the code template name itself you will see the contexts it is permitted.

![image](https://user-images.githubusercontent.com/44065187/202714438-1f42aa51-727d-4279-89ef-ea831b4e8fdc.png)

**Reply by 5011-5011** [reply]
@pacmano1  Perfect. Thank you very much

**Reply by jcurry5** [reply]
> @jcurry5 I add the security library in this section? Or do I install it via console? Thanks.
> 
> ![image](https://user-images.githubusercontent.com/98954172/202709335-de2594c6-274a-4588-99fe-267a795da5ec.png)

I added the jose4j.jar as a resource under Loaded Libraries (under Settings) to utilize the classes that are associated with that jar.

**Reply by pacmano1** [reply]
I think I misunderstood the question - if you are including new jars, that is under "settings"..."resources".   Typically you create a new directory on the server and add your jar there, then create a resource.  Then you include that resource in the channel via the "Set Dependencies" button on the Summary tab of the channel editor.

**Reply by 5011-5011** [reply]
@jcurry5   Could you give me more help? tried to do it but still no success...

**Reply by jcurry5** [reply]
What exactly are you having issues with?

**Comment by siddharthsharma9258** [comment]
If I understand correctly, the user will consume the API on HTTP Listener and request a JWT, and then use that token in subsequent requests until 4 mins validity.

**Reply by jcurry5** [reply]
The destination is set to the HTTP sender end point that sends over the JWT that is created in the transformer.  Then I take the response from the HTTP endpoint and pass that value into the vendor API URL.  The JWT is recycled every 4 minutes per vendor guidelines.

**Comment by bharbhar** [comment, marked as answer]
I encoded the JWT using JavaScript only. Created a JavaScript Writer destination. In that JS Writer destination I wrote the following code. You can copy it and plug in your own payload and keys. This JS Writer destination creates the encoded JWT which is then placed in a channel map variable. Then I grab that variable in the HTTP Sender destination which runs next in query parameters like this: ${jwt}. Then just grab the response token. The HMAC256 code is from CryptoJS and another library for Base64url encoding.

```
//CryptoJS HMAC256 rollup
var CryptoJS=CryptoJS||function(h,s){var f={},g=f.lib={},q=function(){},m=g.Base={extend:function(a){q.prototype=this;var c=new q;a&&c.mixIn(a);c.hasOwnProperty("init")||(c.init=function(){c.$super.init.apply(this,arguments)});c.init.prototype=c;c.$super=this;return c},create:function(){var a=this.extend();a.init.apply(a,arguments);return a},init:function(){},mixIn:function(a){for(var c in a)a.hasOwnProperty(c)&&(this[c]=a[c]);a.hasOwnProperty("toString")&&(this.toString=a.toString)},clone:function(){return this.init.prototype.extend(this)}},
r=g.WordArray=m.extend({init:function(a,c){a=this.words=a||[];this.sigBytes=c!=s?c:4*a.length},toString:function(a){return(a||k).stringify(this)},concat:function(a){var c=this.words,d=a.words,b=this.sigBytes;a=a.sigBytes;this.clamp();if(b%4)for(var e=0;e<a;e++)c[b+e>>>2]|=(d[e>>>2]>>>24-8*(e%4)&255)<<24-8*((b+e)%4);else if(65535<d.length)for(e=0;e<a;e+=4)c[b+e>>>2]=d[e>>>2];else c.push.apply(c,d);this.sigBytes+=a;return this},clamp:function(){var a=this.words,c=this.sigBytes;a[c>>>2]&=4294967295<<
32-8*(c%4);a.length=h.ceil(c/4)},clone:function(){var a=m.clone.call(this);a.words=this.words.slice(0);return a},random:function(a){for(var c=[],d=0;d<a;d+=4)c.push(4294967296*h.random()|0);return new r.init(c,a)}}),l=f.enc={},k=l.Hex={stringify:function(a){var c=a.words;a=a.sigBytes;for(var d=[],b=0;b<a;b++){var e=c[b>>>2]>>>24-8*(b%4)&255;d.push((e>>>4).toString(16));d.push((e&15).toString(16))}return d.join("")},parse:function(a){for(var c=a.length,d=[],b=0;b<c;b+=2)d[b>>>3]|=parseInt(a.substr(b,
2),16)<<24-4*(b%8);return new r.init(d,c/2)}},n=l.Latin1={stringify:function(a){var c=a.words;a=a.sigBytes;for(var d=[],b=0;b<a;b++)d.push(String.fromCharCode(c[b>>>2]>>>24-8*(b%4)&255));return d.join("")},parse:function(a){for(var c=a.length,d=[],b=0;b<c;b++)d[b>>>2]|=(a.charCodeAt(b)&255)<<24-8*(b%4);return new r.init(d,c)}},j=l.Utf8={stringify:function(a){try{return decodeURIComponent(escape(n.stringify(a)))}catch(c){throw Error("Malformed UTF-8 data");}},parse:function(a){return n.parse(unescape(encodeURIComponent(a)))}},
u=g.BufferedBlockAlgorithm=m.extend({reset:function(){this._data=new r.init;this._nDataBytes=0},_append:function(a){"string"==typeof a&&(a=j.parse(a));this._data.concat(a);this._nDataBytes+=a.sigBytes},_process:function(a){var c=this._data,d=c.words,b=c.sigBytes,e=this.blockSize,f=b/(4*e),f=a?h.ceil(f):h.max((f|0)-this._minBufferSize,0);a=f*e;b=h.min(4*a,b);if(a){for(var g=0;g<a;g+=e)this._doProcessBlock(d,g);g=d.splice(0,a);c.sigBytes-=b}return new r.init(g,b)},clone:function(){var a=m.clone.call(this);
a._data=this._data.clone();return a},_minBufferSize:0});g.Hasher=u.extend({cfg:m.extend(),init:function(a){this.cfg=this.cfg.extend(a);this.reset()},reset:function(){u.reset.call(this);this._doReset()},update:function(a){this._append(a);this._process();return this},finalize:function(a){a&&this._append(a);return this._doFinalize()},blockSize:16,_createHelper:function(a){return function(c,d){return(new a.init(d)).finalize(c)}},_createHmacHelper:function(a){return function(c,d){return(new t.HMAC.init(a,
d)).finalize(c)}}});var t=f.algo={};return f}(Math);
(function(h){for(var s=CryptoJS,f=s.lib,g=f.WordArray,q=f.Hasher,f=s.algo,m=[],r=[],l=function(a){return 4294967296*(a-(a|0))|0},k=2,n=0;64>n;){var j;a:{j=k;for(var u=h.sqrt(j),t=2;t<=u;t++)if(!(j%t)){j=!1;break a}j=!0}j&&(8>n&&(m[n]=l(h.pow(k,0.5))),r[n]=l(h.pow(k,1/3)),n++);k++}var a=[],f=f.SHA256=q.extend({_doReset:function(){this._hash=new g.init(m.slice(0))},_doProcessBlock:function(c,d){for(var b=this._hash.words,e=b[0],f=b[1],g=b[2],j=b[3],h=b[4],m=b[5],n=b[6],q=b[7],p=0;64>p;p++){if(16>p)a[p]=
c[d+p]|0;else{var k=a[p-15],l=a[p-2];a[p]=((k<<25|k>>>7)^(k<<14|k>>>18)^k>>>3)+a[p-7]+((l<<15|l>>>17)^(l<<13|l>>>19)^l>>>10)+a[p-16]}k=q+((h<<26|h>>>6)^(h<<21|h>>>11)^(h<<7|h>>>25))+(h&m^~h&n)+r[p]+a[p];l=((e<<30|e>>>2)^(e<<19|e>>>13)^(e<<10|e>>>22))+(e&f^e&g^f&g);q=n;n=m;m=h;h=j+k|0;j=g;g=f;f=e;e=k+l|0}b[0]=b[0]+e|0;b[1]=b[1]+f|0;b[2]=b[2]+g|0;b[3]=b[3]+j|0;b[4]=b[4]+h|0;b[5]=b[5]+m|0;b[6]=b[6]+n|0;b[7]=b[7]+q|0},_doFinalize:function(){var a=this._data,d=a.words,b=8*this._nDataBytes,e=8*a.sigBytes;
d[e>>>5]|=128<<24-e%32;d[(e+64>>>9<<4)+14]=h.floor(b/4294967296);d[(e+64>>>9<<4)+15]=b;a.sigBytes=4*d.length;this._process();return this._hash},clone:function(){var a=q.clone.call(this);a._hash=this._hash.clone();return a}});s.SHA256=q._createHelper(f);s.HmacSHA256=q._createHmacHelper(f)})(Math);
(function(){var h=CryptoJS,s=h.enc.Utf8;h.algo.HMAC=h.lib.Base.extend({init:function(f,g){f=this._hasher=new f.init;"string"==typeof g&&(g=s.parse(g));var h=f.blockSize,m=4*h;g.sigBytes>m&&(g=f.finalize(g));g.clamp();for(var r=this._oKey=g.clone(),l=this._iKey=g.clone(),k=r.words,n=l.words,j=0;j<h;j++)k[j]^=1549556828,n[j]^=909522486;r.sigBytes=l.sigBytes=m;this.reset()},reset:function(){var f=this._hasher;f.reset();f.update(this._iKey)},update:function(f){this._hasher.update(f);return this},finalize:function(f){var g=
this._hasher;f=g.finalize(f);g.reset();return g.finalize(this._oKey.clone().concat(f))}})})();

//CryptoJS Base64 encoding component
(function(){var h=CryptoJS,j=h.lib.WordArray;h.enc.Base64={stringify:function(b){var e=b.words,f=b.sigBytes,c=this._map;b.clamp();b=[];for(var a=0;a<f;a+=3)for(var d=(e[a>>>2]>>>24-8*(a%4)&255)<<16|(e[a+1>>>2]>>>24-8*((a+1)%4)&255)<<8|e[a+2>>>2]>>>24-8*((a+2)%4)&255,g=0;4>g&&a+0.75*g<f;g++)b.push(c.charAt(d>>>6*(3-g)&63));if(e=c.charAt(64))for(;b.length%4;)b.push(e);return b.join("")},parse:function(b){var e=b.length,f=this._map,c=f.charAt(64);c&&(c=b.indexOf(c),-1!=c&&(e=c));for(var c=[],a=0,d=0;d<
e;d++)if(d%4){var g=f.indexOf(b.charAt(d-1))<<2*(d%4),h=f.indexOf(b.charAt(d))>>>6-2*(d%4);c[a>>>2]|=(g|h)<<24-8*(a%4);a++}return j.create(c,a)},_map:"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="}})();

/*
 * JavaScript base64 / base64url encoder and decoder
 */

var b64c = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"   // base64 dictionary
var b64u = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"   // base64url dictionary
var b64pad = ''

/* base64_encode_data
 * Internal helper to encode data to base64 using specified dictionary.
 */
function base64_encode_data(data, len, b64x) {
    var dst = ""
    var i

    for (i = 0; i <= len - 3; i += 3)
    {
        dst += b64x.charAt(data.charCodeAt(i) >>> 2)
        dst += b64x.charAt(((data.charCodeAt(i) & 3) << 4) | (data.charCodeAt(i+1) >>> 4))
        dst += b64x.charAt(((data.charCodeAt(i+1) & 15) << 2) | (data.charCodeAt(i+2) >>> 6))
        dst += b64x.charAt(data.charCodeAt(i+2) & 63)
    }

    if (len % 3 == 2)
    {
        dst += b64x.charAt(data.charCodeAt(i) >>> 2)
        dst += b64x.charAt(((data.charCodeAt(i) & 3) << 4) | (data.charCodeAt(i+1) >>> 4))
        dst += b64x.charAt(((data.charCodeAt(i+1) & 15) << 2))
        dst += b64pad
    }
    else if (len % 3 == 1)
    {
        dst += b64x.charAt(data.charCodeAt(i) >>> 2)
        dst += b64x.charAt(((data.charCodeAt(i) & 3) << 4))
        dst += b64pad
        dst += b64pad
    }

    return dst
}

/* base64_encode
 * Encode a JavaScript string to base64.
 * Specified string is first converted from JavaScript UCS-2 to UTF-8.
 */
function base64_encode(str) {
    var utf8str = unescape(encodeURIComponent(str))
    return base64_encode_data(utf8str, utf8str.length, b64c)
}

/* base64url_encode
 * Encode a JavaScript string to base64url.
 * Specified string is first converted from JavaScript UCS-2 to UTF-8.
 */
function base64url_encode(str) {
    var utf8str = unescape(encodeURIComponent(str))
    return base64_encode_data(utf8str, utf8str.length, b64u)
}

/* base64_charIndex
 * Internal helper to translate a base64 character to its integer index.
 */
function base64_charIndex(c) {
    if (c == "+") return 62
    if (c == "/") return 63
    return b64u.indexOf(c)
}

/* base64_decode
 * Decode a base64 or base64url string to a JavaScript string.
 * Input is assumed to be a base64/base64url encoded UTF-8 string.
 * Returned result is a JavaScript (UCS-2) string.
 */
function base64_decode(data) {
    var dst = ""
    var i, a, b, c, d, z
    
    for (i = 0; i < data.length - 3; i += 4) {
        a = base64_charIndex(data.charAt(i+0))
        b = base64_charIndex(data.charAt(i+1))
        c = base64_charIndex(data.charAt(i+2))
        d = base64_charIndex(data.charAt(i+3))

        dst += String.fromCharCode((a << 2) | (b >>> 4))
        if (data.charAt(i+2) != b64pad)
            dst += String.fromCharCode(((b << 4) & 0xF0) | ((c >>> 2) & 0x0F))
        if (data.charAt(i+3) != b64pad)
            dst += String.fromCharCode(((c << 6) & 0xC0) | d)
    }

    dst = decodeURIComponent(escape(dst))
    return dst
}

/* base64url_sniff
 * Check whether specified base64 string contains base64url specific characters.
 * Return true if specified string is base64url encoded, false otherwise.
 */
function base64url_sniff(base64) {
    if (base64.indexOf("-") >= 0) return true
    if (base64.indexOf("_") >= 0) return true
    return false
}


//JWT Encoding

var headerJSON = {
  "alg": "HS256",
  "typ": "JWT"
};

var encodedHeader = base64url_encode(JSON.stringify(headerJSON));

const iat = Math.floor(Date.now() / 1000);

logger.info("Current iat " + iat);

var dataJSON = {
  "sub": "03f339da-81f5-4267-bca9-0ea5b4779557",
  "iat": iat,
  "exp": iat+600
}

var encodedData = base64url_encode(JSON.stringify(dataJSON));

var token = encodedHeader + "." + encodedData;

var secret = "44008c56-5361-432c-bf88-26625ddfc407";

var signature = CryptoJS.HmacSHA256(token, secret);

signature = CryptoJS.enc.Base64.stringify(signature);

signature = signature.replace('=','').replace('/','_').replace('+','-');

var signedToken = token + "." + signature;

logger.info("Token " + token);
logger.info("Signature " + signature);
logger.info("Signed Token " + signedToken);

channelMap.put('jwt', signedToken);
```

**Reply by 5011-5011** [reply]
Thank you very much for sharing it with me. I'll try.

**Reply by 5011-5011** [reply]
@bharbhar Could you help me, maybe I'm wrong.
I managed to create the jwt, however its expiration is giving me problems. I have 1 channel that fulfills the function of generating the jwt, I save it in a global variable to include this variable in a filter in the following channels.
When I send the request to request the token, it returns it to me without a problem, but the problem is that it generates a new one for each request it receives.
If you could guide me how you applied the logic, I would really appreciate it.

![image](https://user-images.githubusercontent.com/98954172/221872353-569e6150-82c1-49f2-8c70-6963410949cf.png)

**Reply by bharbhar** [reply]
That is correct this code will fire and make a new token each time. I implemented as a destination and channel variable only. Each message would fire off, get a new token and process that message using the new token. This channel a volume of hundreds of messages per day. The old tokens should still remain valid even if you get a new token. Maybe try to change your design and move things around.

You can control expiration of these tokens by increasing iat.

**Reply by 5011-5011** [reply]
@bharbhar hello! i would like to know if you can give me some guidance. 
I'm working on filtering the jwt according to their time of broadcast. I have tried to do this by creating functions that take the jwt from the header of the incoming request, then decode it and filter it by its time of emission. 
However it has not been successful, I think the problem is because it does not decode the jwt and the following is not processed. 
I have read that certain libraries like jsonwebtoken are not compatible with Mirth connect, I have my doubts about this. 
If you could help me I would appreciate it, I've been stuck on this for a while.

**Reply by pacmano1** [reply]
@5011-5011 Please don't hijack the thread with a new topic, can you start a new thread referencing this one if needed?

**Reply by 5011-5011** [reply]
@pacmano1 Is done.

**Comment by twest-mirthconnect** [comment]
@bharbhar -  Can I suggest a PR for this topic? We have reviewed the conversation and would like to include this in a future version of MC.

**Reply by narupley** [reply]
Specifically, a PR for https://github.com/nextgenhealthcare/connect-examples

Like perhaps encapsulate this all within a Code Template (also maybe an example channel using said code template) that anyone can import and use!

**Reply by bharbhar** [reply]
How flattering. Let me take a look and try to put this together.

**Reply by ds2-jetten** [reply]
> Like perhaps encapsulate this all within a Code Template (also maybe an example channel using said code template) that anyone can import and use!

I second this!

**Comment by ds2-jetten** [comment]
@bharbhar How did you go about creating the CryptoJS rollup? I have a similar integration I am working on, but I need to use sha256.

**Reply by bharbhar** [reply]
The roll-ups are provided in the CryptoJS library which I got from here: https://code.google.com/archive/p/crypto-js/downloads. I ended up having to roll it up myself because none of the roll-ups worked out of the box with base64url encoding. So I just hand picked them and plugged in another library for the url part of it. I see the v 3.1.2 has sha256 as a component and a roll-up.

**Comment by pacmano1** [comment]
Here is another way:

https://github.com/pacmano1/DiridiumTechnologies/blob/main/epicJWTforBackend.js
